### PR TITLE
Week5

### PR DESCRIPTION
**Pull request by oishikg**
Would be helpful if you guys could answer some of the questions I ask in the comments!

### CODE REVIEW

**Review by arbus — APPROVED** [review]

File: week5/hw5/MyLib.hs [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+  &&
+  (f e5 == 0) 
+  
+{- function to evaluate the expression} -} 
+eval :: ExprT -> Integer
+eval exp =
+  case exp of
+    Lit n -> n
+    Add e1 e2 -> (eval e1) + (eval e2)
+    Mul e1 e2 -> (eval e1) * (eval e2)
```
Comment (JolandaNava) [inline week5/hw5/MyLib.hs]: You don't even need brackets here :)

File: week5/hw5/MyLib.hs [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+class Expr t where
+  lit :: Integer -> t
+  add :: t -> t -> t
+  mul :: t -> t -> t
+
+{- ExprT as an instance of the type class -} 
+instance Expr ExprT where
+  lit e = Lit e
+  add e1 e2 = Add e1 e2
+  mul e1 e2 = Mul e1 e2
```
Comment (JolandaNava) [inline week5/hw5/MyLib.hs]: because `Add` is a function that takes two arguments, and because the arguments that `add` takes are the same, you can just write:
```
    lit = Lit
    add = Add
    mul = Mul
```

File: week5/hw5/MyLib.hs [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+
+Try printing out each of those tests in ghci to see if things are
+working. It’s great how easy it is for us to swap in new semantics for
+the same syntactic expression! -}
+
+{- Integers as an instance of the Expr type class -}
+instance Expr Integer where
+  lit n = n
+  add n1 n2 = n1 + n2
+  mul n1 n2 = n1 * n2
```
Comment (JolandaNava) [inline week5/hw5/MyLib.hs]: what's the function for which `fun x = x`? 

if you want to be even drier here, you can use the prefix version of the operators, e.g.: `add = (+)` since `1+2` and `(+) 1 2` are equivalent

File: week5/hw5/MyLib.hs [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+
+{- Booleans as an instance of the Expr type class -}
+instance Expr Bool where
+  lit n
+    | n <= 0 = False
+    | otherwise = True
```
Comment (JolandaNava) [inline week5/hw5/MyLib.hs]: `n <= 0` already produces a `Bool` value, so you can shorten this to `lit n = n <=0`, or go wild with prefix notation and say `lit = (<= 0)` (same comment on the prefix notation goes for the operators below)

File: week5/hw5/MyLib.hs [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+
+  add bv1 bv2 = bv1 || bv2
+  mul bv1 bv2 = bv1 && bv2 
+  
+{- Defining wrappers for MinMax -}
+newtype MinMax = MinMax Integer deriving (Eq, Show)
+
+instance Expr MinMax where
+  lit n = MinMax n
```
Comment (JolandaNava) [inline week5/hw5/MyLib.hs]: you can drop the `n`

File: week5/hw5/MyLib.hs [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+  add (MinMax n1) (MinMax n2) = MinMax (max n1 n2)
+  mul (MinMax n1) (MinMax n2) = MinMax (min n1 n2)
```
Comment (JolandaNava) [inline week5/hw5/MyLib.hs]: You can start using the `$` operator to reduce the number of brackets. (Definition [here](https://hackage.haskell.org/package/hspec-2.7.1/docs/Test-Hspec-Discover.html#v:-36-) )

Basically it means, evaluate what's to my right, then pass it to the function on my left. 

example: `MinMax $ max n1 n2`

File: week5/hw5/MyLib.hs [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+
+
+{- Defining wrappers for Mod7 -}
+newtype Mod7 = Mod7 Integer deriving (Eq, Show)
+
+instance Expr Mod7 where
+  lit n = Mod7 (n `mod` 7)
+  add (Mod7 n1) (Mod7 n2) = Mod7 ((n1 + n2) `mod` 7)
+  mul (Mod7 n1) (Mod7 n2) = Mod7 ((n1 * n2) `mod` 7)
```
Comment (JolandaNava) [inline week5/hw5/MyLib.hs]: See if you can use the `$` operator here too! Also, you can use `lit` in your definition of `add` and `mul` if it's helpful.

File: week5/hw5/MyLib.hs [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+  mul p1 p2 = p2 ++ p1 ++ [SVM.Mul]
+
+{- Here's the fun part: since the parser parses the input string based on the
+semantics of the caller, we simply provide it with the semantics for the program type.
+Note that the type definition of compile the type-inferencer to infer that the
+lit, add, and mul operations to be used are those defined for the Program instance
+-}
+
+-- Question: How does Haskell infer that compile returns a Maybe SVM.Program even when
+-- the type defintion is commented out?
```
Comment (JolandaNava) [inline week5/hw5/MyLib.hs:363]: Haskell looks at where the function is used to infer type constraints: in this file you use `compile` in your `test` function. In that function it knows that `compile str` returns a `Maybe a` and then it knows that the `a` is given as argument to `SVM.stackVM`, which takes a `SVM.Program` - so the type signature of `compile` has to be `str -> Maybe SVM.Program`

File: week5/hw5/MyLib.hs [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+
+-- Note that the type of parseExp is :
+-- parseExp :: (Integer -> a) (a -> a -> a) -> (a -> a -> a) -> a
+-- That is, the parser parses the string based on the semantics supplied buy the user
+
+evalStr :: String -> Maybe Integer
+evalStr = (\x ->
+             case x of
+               Nothing -> Nothing
+               Just e -> Just (eval e)) . (parseExp Lit Add Mul)
```
Comment (JolandaNava) [inline week5/hw5/MyLib.hs:126]: Not necessary, but I think you might find this interesting (and another step into the world of Haskell operators): [`<$>`](https://hackage.haskell.org/package/base-4.12.0.0/docs/Prelude.html#v:-60--36--62-)

```
(<$>) :: Functor f => (a -> b) -> f a -> f b 
```
not sure how much you've read on Functors so far, but for now you can think of this in terms of `Maybe`:

```
(<$>) :: (a -> b) -> Maybe a -> Maybe b 
```

So if I have a function from `a` to `b` and I have a `Maybe a` I can make a `Maybe b` by doing pretty much what you wrote above
```
case x of
    Nothing -> Nothing
    Just a -> Just $ fun a
```

See if you can rewrite `evalStr` using the `<$>` operator instead of matching on the element explicitely

File: week5/hw5/MyLib.hs (hunk 1 above)
Comment (oishikg) [reply week5/hw5/MyLib.hs]: Noted, since function application has precedence over addition.

File: week5/hw5/MyLib.hs (hunk 3 above)
Comment (oishikg) [reply week5/hw5/MyLib.hs]: Yeah it's the `id` function! Added these changes!

File: week5/hw5/MyLib.hs (hunk 4 above)
Comment (oishikg) [reply week5/hw5/MyLib.hs]: Yes, it's basically `lit = (> 0)`. Foolish of me to compute a `Bool` and then invert it!

File: week5/hw5/MyLib.hs (hunk 6 above)
Comment (oishikg) [reply week5/hw5/MyLib.hs]: Indeed, a very handy operator!

File: week5/hw5/MyLib.hs (hunk 7 above)
Comment (oishikg) [reply week5/hw5/MyLib.hs]: Of course it's `lit $ n1 + n2` (for instance). Implemented this change.

File: week5/hw5/MyLib.hs (hunk 8 above)
Comment (oishikg) [reply week5/hw5/MyLib.hs:363]: Thanks, this explanation makes a lot of sense!

File: week5/hw5/MyLib.hs (hunk 9 above)
Comment (oishikg) [reply week5/hw5/MyLib.hs:126]: Moving on to functors soon! Will start incorporating them.

**Review by JolandaNava — APPROVED** [review]
